### ISSUE TITLE
Electron 5+ Node 12 rebuild error

### ISSUE THREAD

**Issue by daCapricorn** [theturtle32/WebSocket-Node#358]
An unhandled error occurred inside electron-rebuild
gyp info it worked if it ends with ok
gyp info using node-gyp@5.0.2
gyp info using node@12.0.0 | darwin | x64
gyp info find Python using Python version 2.7.10 found at "/usr/bin/python"
gyp info spawn /usr/bin/python
gyp info spawn args [
gyp info spawn args   '/Users/steven/Project/Node/ScatterDesktop/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/steven/Project/Node/ScatterDesktop/node_modules/websocket/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/steven/Project/Node/ScatterDesktop/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/steven/.electron-gyp/5.0.1/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/steven/.electron-gyp/5.0.1',
gyp info spawn args   '-Dnode_gyp_dir=/Users/steven/Project/Node/ScatterDesktop/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/steven/.electron-gyp/5.0.1/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/steven/Project/Node/ScatterDesktop/node_modules/websocket',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CXX(target) Release/obj.target/bufferutil/src/bufferutil.o
../src/bufferutil.cc:26:30: error: no template named 'Handle' in namespace 'v8'
  static void Initialize(v8::Handle<v8::Object> target)
                         ~~~~^
../src/bufferutil.cc:34:74: warning: 'GetFunction' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Nan::Set(target, Nan::New<String>("BufferUtil").ToLocalChecked(), t->GetFunction());
                                                                         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:5881:3: note: 'GetFunction' has been explicitly marked deprecated here
  V8_DEPRECATED("Use maybe version", Local<Function> GetFunction());
  ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:50:40: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<Object> bufferObj = info[0]->ToObject();
                                       ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:57:34: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
      Local<Object> src = array->Get(i)->ToObject();
                                 ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:3317:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(uint32_t index));
  ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:57:42: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
      Local<Object> src = array->Get(i)->ToObject();
                                         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:68:41: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<Object> buffer_obj = info[0]->ToObject();
                                        ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:70:39: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<Object> mask_obj = info[1]->ToObject();
                                      ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:89:41: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<Object> buffer_obj = info[0]->ToObject();
                                        ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:90:39: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<Object> mask_obj = info[1]->ToObject();
                                      ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:92:41: warning: 'ToObject' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<Object> output_obj = info[2]->ToObject();
                                        ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2419:10: note: 'ToObject' has been explicitly marked deprecated here
  inline V8_DEPRECATED("Use maybe version", Local<Object> ToObject() const);
         ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:93:40: warning: 'Int32Value' is deprecated: Use maybe version [-Wdeprecated-declarations]
    unsigned int dataOffset = info[3]->Int32Value();
                                       ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2446:3: note: 'Int32Value' has been explicitly marked deprecated here
  V8_DEPRECATED("Use maybe version", int32_t Int32Value() const);
  ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:94:36: warning: 'Int32Value' is deprecated: Use maybe version [-Wdeprecated-declarations]
    unsigned int length = info[4]->Int32Value();
                                   ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:2446:3: note: 'Int32Value' has been explicitly marked deprecated here
  V8_DEPRECATED("Use maybe version", int32_t Int32Value() const);
  ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../src/bufferutil.cc:115:6: error: variable has incomplete type 'void'
void init (Handle<Object> target)
     ^
../src/bufferutil.cc:115:19: error: 'Object' does not refer to a value
void init (Handle<Object> target)
                  ^
/Users/steven/.electron-gyp/5.0.1/include/node/v8.h:3257:17: note: declared here
class V8_EXPORT Object : public Value {
                ^
../src/bufferutil.cc:115:12: error: use of undeclared identifier 'Handle'
void init (Handle<Object> target)
           ^
../src/bufferutil.cc:115:27: error: use of undeclared identifier 'target'
void init (Handle<Object> target)
                          ^
../src/bufferutil.cc:115:34: error: expected ';' after top level declarator
void init (Handle<Object> target)
                                 ^
                                 ;
11 warnings and 6 errors generated.
make: *** [Release/obj.target/bufferutil/src/bufferutil.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/steven/Project/Node/ScatterDesktop/node_modules/node-gyp/lib/build.js:190:23)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:256:12)
gyp ERR! System Darwin 18.6.0
gyp ERR! command "/usr/local/bin/node" "/Users/steven/Project/Node/ScatterDesktop/node_modules/electron-rebuild/node_modules/.bin/node-gyp" "rebuild" "--target=5.0.1" "--arch=x64" "--dist-url=https://electronjs.org/headers" "--build-from-source"
gyp ERR! cwd /Users/steven/Project/Node/ScatterDesktop/node_modules/websocket
gyp ERR! node -v v12.0.0
gyp ERR! node-gyp -v v5.0.2
gyp ERR! not ok

Failed with exit code: 1

**Comment by ibc** [comment]
Can you specify a bit more the error (instead of just pasting the raw console output)?

I can perfectly run `npm install websocket` using Node 12.6.0.

**Comment by daCapricorn** [comment]
@ibc To use native modules in electron, modules should be rebuilt by using electron-rebuild. I create a demo project, you can test [it](https://github.com/daCapricorn/electron-quick-start). Just run `npm install`.

**Comment by daCapricorn** [comment]
https://github.com/electron/electron/blob/master/docs/tutorial/using-native-node-modules.md

**Comment by ibc** [comment]
Too hard for me, sorry. Zero experience with Electron.

**Comment by daCapricorn** [comment]
@ibc yanyiwu/nodejieba#132 may be helpful. I think should use `Local` instead of `Handle` in bufferutil.cc and validation.cc.

**Comment by ibc** [comment]
I've not done this and I have zero experience with Node native modules. Would you create a PR?

**Comment by daCapricorn** [comment]
I will try it. nodejs/nan#849

**Comment by daCapricorn** [comment]
@ibc done theturtle32/WebSocket-Node#361

**Comment by belcherj** [comment]
I am still getting an error:
```
[electron] An unhandled error occurred inside electron-rebuild
[electron]   CXX(target) Release/obj.target/bufferutil/src/bufferutil.o
[electron] ../src/bufferutil.cc:34:86: error: too few arguments to function call, single argument 'context' was not specified
[electron]     Nan::Set(target, Nan::New<String>("BufferUtil").ToLocalChecked(), t->GetFunction());
[electron]                                                                       ~~~~~~~~~~~~~~ ^
```

**Comment by belcherj** [comment]
You can reproduce by pulling this PR and running `npm install && npm run dev`.  It has build from master of this repo to test #361.  https://github.com/Automattic/simplenote-electron/pull/1496

**Comment by Dirk-** [comment]
This is still a problem with electron 6.0.11. See [my comment here](https://github.com/theturtle32/WebSocket-Node/issues/369#issuecomment-537949699).

**Comment by theturtle32** [comment]
This should be fixed in v1.0.31 released today
